lib/srv/transport/transportv1/transport.go

```diff
@@ -125,3 +149,128 @@ func (c clusterStream) Recv() ([]byte, error) {
 func (c clusterStream) Send(frame []byte) error {
 	return trace.Wrap(c.stream.Send(&proxyv1.ProxyClusterResponse{Frame: &proxyv1.Frame{Payload: frame}}))
 }
+
+// ProxySSH establishes a connection to a host and proxies both the SSH and SSH
+// Agent protocol over the stream. The first request from the client must contain
+// a valid dial target before the connection can be established.
+func (s *Service) ProxySSH(stream proxyv1.ProxyService_ProxySSHServer) error {
+	ctx := stream.Context()
+
+	// wait for the first request to arrive with the dial request
+	req, err := stream.Recv()
+	if err != nil {
+		return trace.Wrap(err, "failed receiving first frame")
+	}
+
+	// validate the target
+	if req.DialTarget == nil {
+		return trace.BadParameter("first frame must contain a dial target")
+	}
+
+	host, port, err := net.SplitHostPort(req.DialTarget.HostPort)
+	if err != nil {
+		return trace.BadParameter("dial target contains an invalid hostport")
+	}
+
+	// create a stream for SSH Agent protocol
+	agentStream := newSSHStream(stream, func(payload []byte) *proxyv1.ProxySSHResponse {
+		return &proxyv1.ProxySSHResponse{Frame: &proxyv1.ProxySSHResponse_Agent{Agent: &proxyv1.Frame{Payload: payload}}}
+	})
+
+	// create a stream for SSH protocol
+	sshStream := newSSHStream(stream, func(payload []byte) *proxyv1.ProxySSHResponse {
+		return &proxyv1.ProxySSHResponse{Frame: &proxyv1.ProxySSHResponse_Ssh{Ssh: &proxyv1.Frame{Payload: payload}}}
+	})
+
+	// multiplex incoming frames to the appropriate protocol
+	// handlers for the duration of the stream
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+
+				s.cfg.Logger.Errorf("ssh stream terminated unexpectedly: %v", err)
+				return
+			}
+
+			switch frame := req.Frame.(type) {
+			case *proxyv1.ProxySSHRequest_Ssh:
+				sshStream.incomingC <- frame.Ssh.Payload
+			case *proxyv1.ProxySSHRequest_Agent:
+				agentStream.incomingC <- frame.Agent.Payload
+			default:
+				s.cfg.Logger.Errorf("received unexpected ssh frame: %T", frame)
+				continue
+			}
```

Comment:
Nit: I'm guessing these sends don't select on `stream.Context().Done()` because we want to ensure that all received messages are consumed, even if the stream is closing... assuming that's the case, maybe worth adding comments to `sshStream.Recv` indicating that failure to drain could result in a leak?